Add wrapper for pubtools-marketplacesvm

JAVGan opened the pull request:
This commit introduces a wrapper for `pubtools-marketplacesvm` which aims to be eventually used to support VM images pushes on various cloudmarketplaces (such as AWS and Azure).

With this wrapper it will be possible to easily integrate the tooling into Tekton CI using the Cloud Staged structure as input.

Refers to SPSTRAT-464

ralphbean commented:
I defer to @release-service-maintainers on this.

mmalina commented on pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper.py:
First of all, if you want this in the image, you would need to modify the Dockerfile. But I wonder if we need a new directory for each new wrapper, especially since we already have one pulp related wrapper. So maybe rename the existing directory to `pubtools` or `pubtools-wrappers` and have both scripts in it?

JAVGan replied:
Thanks for letting me know! I'll add the tooling to the `Dockerfile`.

About having a common directory for every `pubtools` I'm not sure, it's up to you... They do different things but they were both used in Pub to release content to different targets.

johnbieren replied:
I'm +1 to one `pubtools` dir

mmalina commented on Dockerfile:
```diff
@@ -60,6 +61,7 @@ COPY pyxis /home/pyxis
 COPY utils /home/utils
 COPY templates /home/templates
 COPY pubtools-pulp-wrapper /home/pubtools-pulp-wrapper
+COPY pubtools-marketplacesvm-wrapper /home/pubtools-marketplacesvm-wrapper
```
Thanks! @johnbieren WDYT? It just seems like too much clutter to have one dir for each wrapper script. Maybe we should move them all to a new `wrappers` directory? It should have zero impact, because typically you would rely on PATH when calling them (but of course better to check in the catalog and internal repos).

johnbieren replied:
I am fine with a `wrappers` dir or at minimum a `pubtools` dir. If we want to go with `wrappers` we can leave this PR as is and move them all in a followup PR. Just let me know what you prefer as far as the organization

mmalina replied:
Yeah, I would go with wrappers and move all 3 files in it. @JAVGan let us know if you're fine doing it here, or you'd prefer us to do it separately.

JAVGan replied:
I prefer to have it in a separate MR with your assistance to avoid breaking things as well as mixing the implementation of a new tooling wrapper with the refactoring.

mmalina replied:
ok

johnbieren commented on pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper.py:
```diff
@@ -0,0 +1,149 @@
+For more information please refer to documentation:
+* https://github.com/release-engineering/pubtools-marketplacesvm
+* https://release-engineering.github.io/pushsource/sources/staged.html#root-destination-cloud-images
+* https://release-engineering.github.io/pushsource/schema/cloud.html#cloud-schema
```
If there is a slack channel your team hangs out on, it would be great to also add it here

JAVGan replied:
I've added our `#stratosphere-bau` on it, thanks for the suggestion!

Also I've ran `black -S .` again to reformat the scripts

johnbieren commented on pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper.py:
```diff
@@ -0,0 +1,149 @@
+
+
+if __name__ == "__main__":
+    entrypoint()
```
Any reason why not to just put `main()` here and not `def entrypoint():` a few lines up?

JAVGan replied:
I was just following the same pattern I found in the other wrapper, but yes, it's not necessary, I'll simplify it

JAVGan replied:
Done simplifying it

johnbieren commented on pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper.py:
```diff
@@ -0,0 +1,149 @@
+    return parser.parse_args()
+
+
+def get_source_url(stagedirs):
```
worth adding a unit test for this?

JAVGan replied:
I've added the tests, PTAL again

johnbieren reviewed (commented):
You are failing for black formatting

lslebodn commented on pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper.py:
```diff
@@ -0,0 +1,148 @@
+        if not re.match(r"^/[^,]{0,4000}\/*starmap\/*$", item):
+            raise ValueError("Not a valid staging directory: %s" % item)
+
+    return f"staged:{','.join(stagedirs)}"
```
Do we plan to validate staging structure with `pushsource-ls` here or it will be done in tekton task?
I am fine either way

JAVGan replied:
I think it makes sense to do that in the Task itself.

johnbieren reviewed (commented):
You're failing flake8 now for some lines being too long. You are also failing gitlint for your commit title. Maybe something like `feat(SPSTRAT-464): add wrapper for pubtools-marketplacesvm` instead? Otherwise lgtm

JAVGan commented:
> You're failing flake8 now for some lines being too long. You are also failing gitlint for your commit title. Maybe something like `feat(SPSTRAT-464): add wrapper for pubtools-marketplacesvm` instead? Otherwise lgtm

@johnbieren I've fixed most of the issues with `flake` except the URL which I've added a `# noqa` since it can't be shortened.

For the commit name I was a bit afraid it woul bee too long to put what you suggested (and fail again on gitlinkt) then I've shortened `pubtools-marketplacesvm` just to `marketplacesvm`.. Hope it works for you

johnbieren commented:
Failing gitlint now for `1: CC1 Body does not contain a 'Signed-off-by' line` `git commit -s` should do that for you. Python checks pass now, great :slightly_smiling_face: 
Can you fix the gitlint thing and rebase? Then I will trigger the e2e and approve. If it looks good to Martin in his morning he can merge

JAVGan commented:
Done rebasing and signing it off.

@mmalina please check everything and merge if it's ok to you 😄

johnbieren commented:
/ok-to-test

JAVGan commented:
@mmalina I noticed some small nitpiks while writing the Task:

1. The `starmap` dir should be inside the staged dir, it's not part of its base directory, thus I'll remove it from here and validate it in the task.
2. There's one missing argument for our tooling, called `--offline` which is meant to prevent the tooling to invoke the StArMap server when the `--repo-file` is set.

I'll create an additional commit with such fixes, then you may squash them on merging if you prefer 😄 

Just a minute, working on that

JAVGan commented:
@johnbieren @mmalina I've added the necessary fixes on bdb3b59c01fbed3bde9a3bd5ac09f7ac4c72040c

Feel free to squash it when merging and please re-run the checks + approval if ok to you

Sorry for the inconvenience

mmalina commented:
@JAVGan sorry for the delay. Can you rebase and squash the commits yourself? The second one fails gitlint now because of capital letter after `feat:`. It looks good to me now. But Let's wait for Johnny as well (should be here when US wakes up).

johnbieren commented:
/ok-to-test
